[#59, #70] Make PR and branch name configurable

pmeinhardt opened the pull request:
Note: For additional details, see the commit messages.

References:
- [Safari Extensions Development Guide "Settings and Local Storage"](https://developer.apple.com/library/content/documentation/Tools/Conceptual/SafariExtensionGuide/ExtensionSettings/ExtensionSettings.html)
- [MDN web docs: Add-ons: "Implement a settings page"](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Implement_a_settings_page)

Templating:
- ~~[Using eval in Chrome Extensions. Safely.](https://developer.chrome.com/extensions/sandboxingEval) – `eval`/`Function` is used in most templating libs we'd use for generating the branch names/commit messages~~
- Templating with regex

TODOs

- [x] Implement preferences in Safari
- [x] Implement preferences in Chrome, Firefox…
- [x] ~~Get templates to work with the CSP restrictions~~
- [x] Format branches, commits and commands according to templates
- [x] Add more tests for `Form` and `TemplateInput`
- [x] Improve options page styling
- [x] Factor out `enhance` function to attach `fmt` to input tickets
- [x] Update README ("Advanced" section with template options)

andreasknoepfle commented on package.json:
```diff
@@ -18,7 +18,7 @@
     "watch:firefox-local": "npm run build:firefox-local -- --watch",
     "watch:safari": "npm run build:safari -- --watch",
     "open:chrome": "./script/open-in-chrome ./dist/chrome https://github.com/bitcrowd/tickety-tick",
-    "open:firefox": "web-ext run --source-dir ./dist/firefox --pref startup.homepage_welcome_url=https://github.com/bitcrowd/tickety-tick",
```
Sadly we need an addon-id to locally test out firefox now. So I switched the `open:firefox` task to `open:firefox-local` since the extensions breaks when doing `open:firefox`.

andreasknoepfle replied:
https://developer.mozilla.org/en-US/Add-ons/WebExtensions/WebExtensions_and_the_Add-on_ID#When_do_you_need_an_add-on_ID

pmeinhardt commented:
@andreasknoepfle, I've updated the PR (squashed & rebased, so please be sure to update your local branch):

- Extracted the template compilation and added separate tests for it
- Removed unused sandbox code that was still lurking around
- Added tests for the new template helpers we added (plus removed `dasherize` for now, since we have `slugify` which is more useful as it also deals with non-ASCII characters and can be used for the same purpose)
- Added a few first tests for `Form` and `TemplateInput` (more needed)

I've updated the TODO list in the PR description to keep track of what we still need to do.

pmeinhardt commented:
@bitboxer and @Empty2k12, this PR implements the enhancements you requested in #59 and #70:
*Branch names*, *commit messages* and even the generated *commands* are now customizable.

We still want to add a couple of additional tests and update the documentation, but you can already start using the new functionality if you build from this branch. We'd be very happy about your feedback so we can address any remaining details 🙌

bitboxer commented:
👍 thanks for this. Currently I am in "newborn needs food" mode and won't be able to test it before the next year. Will try to do it then.

pmeinhardt commented:
👋 Yay @bitboxer, congrats and all the best!!!

pmeinhardt commented:
Completed tests for the option page components.

I've also updated the options page styling and fixed the "ExtractTextPlugin" filename template to automatically name the extracted CSS file after the entry point (`[name].css`). Like this, we generate separate `popup.css` and `options.css` files which are injected into the respective HTML pages.

Here's what the options look like in Chrome and Firefox:

![Options in Chrome](https://user-images.githubusercontent.com/706519/34281260-5d936cae-e6b5-11e7-9ec0-49d46b27118f.png)

![Options in Firefox](https://user-images.githubusercontent.com/706519/34281279-7ea622ce-e6b5-11e7-8fa7-d17b3c9a4d21.png)

Last things to do are:

- Update the README
- Have people test it

Feedback very welcome 🙂

pmeinhardt commented:
Ah, I almost forgot: This what the options look like in Safari:

![Options in Safari](https://user-images.githubusercontent.com/706519/34281387-4464e69e-e6b6-11e7-91f4-cf661ac54e8d.png)

bitboxer commented:
Awesome. Finally found some time to play with it. Works like a charm 👍 . 

![Bowing Wanes World](https://media.giphy.com/media/3o7TKqFZJpwL8Giz2E/giphy.gif)

pmeinhardt commented:
That's so great! Thanks a lot for test-driving it @bitboxer, and a happy new year to you 👋 
I'll update the readme so we can merge and release a new version as soon as there's time.

pmeinhardt commented on src/safari-extension/popup/popup.js:
```diff
@@ -38,7 +40,28 @@ function onPopover(event) {
+        command: fmt.command(ticket),
+      },
+      ...ticket,
+    });
```
This should be factored out because it's the same for both Safari and WebExtension browsers.

pmeinhardt commented:
@andreasknoepfle, when you have time, please take a look at the last four commits I've added.

If everything's okay, feel free to merge and ship :shipit: 

These issues can then be closed:

- #59 
- #70 

I've created an [issue](https://github.com/bitcrowd/tickety-tick/issues/87) for you, for pushing out version 2.0 to the Mozilla and Chrome extension stores.

🙌

pmeinhardt commented:
Here's the milestone for an overview: https://github.com/bitcrowd/tickety-tick/milestones/2.0
